firebase-firestore/src/main/java/com/google/firebase/firestore/core/UserData.java

```diff
@@ -303,9 +301,17 @@ private void validatePath() {
       }
     }
 
+    private static final String RESERVED = "__";
+
     private void validatePathSegment(String segment) {
-      if (isWrite() && reservedFieldRegex.matcher(segment).find()) {
-        throw this.createError("Document fields cannot begin and end with __");
+      if (!isWrite()) return;
+
+      if (segment.isEmpty()) {
+        throw this.createError("Document fields must not be empty");
+      }
+
+      if (segment.startsWith(RESERVED) || segment.endsWith(RESERVED)) {
```

Comment:
I don't understand this change...  Today the clients and backend allow `__x` and `x__` but not `__x__`.  Why are we making the client more restrictive?